_posts/2020-02-26-demo-fraud-detection-2.md

```diff
@@ -0,0 +1,210 @@
+
+Let's first have a look at the [previously-defined](https://flink.apache.org/news/2020/01/15/demo-fraud-detection.html#dynamic-data-partitioning) data-processing pipeline:
+
+```java
+DataStream<Alert> alerts =
+    transactions
+        .process(new DynamicKeyFunction())
+        .keyBy((keyed) -> keyed.getKey());
+        .process(new DynamicAlertFunction())
+```
+
+`DynamicKeyFunction` provides dynamic data partitioning while `DynamicAlertFunction` is responsible for executing the main logic of processing transactions and sending alert messages according to defined rules.
+
+Vol.1 of this case study simplified the use case and assumed that the applied set of rules is pre-initialized and is accessible via the `List<Rules>` within `DynamicKeyFunction`.
+
+```java
+public class DynamicKeyFunction
+    extends ProcessFunction<Transaction, Keyed<Transaction, String, Integer>> {
+
+  /* Simplified */
+  List<Rule> rules = /* Rules that are initialized somehow.*/;
+  ...
+}
+```
+
+Adding rules to this list is obviously possible directly inside the code of the Flink Job at the stage of its initialization (Create a `List` object; use it's `add` method). A major drawback of doing so is that it will require recompilation of the job with each rule modification. In a real Fraud Detection system, rules are expected to change on a frequent basis, and makes this approach unacceptable from the point of view of business and operational requirements. A different approach is needed.
+
+Let's next have a look at a sample rule definition that we introduced in the previous post:
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/patterns-blog-2/rule-dsl.png" width="800px" alt="Figure 1: Rule definition"/>
+<br/>
+<i><small>Figure 1: Rule definition</small></i>
+</center>
+<br/>
+
+![](./../img/blog/patterns-blog-2/rule-dsl.png)
+The previous post covered use of `groupingKeyNames` by `DynamicKeyFunction` to extract message keys. Parameters from the second part of this rule are used by `DynamicAlertFunction`: they define the actual logic of the performed operations and their parameters (such as the alert-triggering limit). This means that the same rule must be present in both `DynamicKeyFunction` and `DynamicAlertFunction`. To achieve this result, we will use the [broadcast data distribution mechanism](https://ci.apache.org/projects/flink/flink-docs-release-1.10/dev/stream/state/broadcast_state.html) of Apache Flink.
+
+Figure 2 presents the final job graph of the system that we are building:
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/patterns-blog-2/job-graph.png" width="800px" alt="Figure 2: Job Graph of the Fraud Detection Flink Job"/>
+<br/>
+<i><small>Figure 2: Job Graph of the Fraud Detection Flink Job</small></i>
+</center>
+<br/>
+
+![](./../img/blog/patterns-blog-2/job-graph.png)
+
+The main blocks of the Transactions processing pipeline are:<br>
+
+* **Transaction Source**, consuming transaction messages from Kafka partitions in parallel. <br>
```

Comment:
```suggestion
* **Transaction Source** that consumes transaction messages from Kafka partitions in parallel. <br>
```

Reply:
I wanted to avoid using "X that ..." in all three bullet points. Is my version hard to understand?

Reply:
this is more because it's not consistent, the bullet points have 

- **Transaction Source**, consuming
- **Dynamic Key Function** that performs data....
- **Dynamic Alert Function** that accumulates a....